presto-common/src/main/java/com/facebook/presto/common/zorder/ZOrderIntegers.java

```diff
@@ -0,0 +1,167 @@
+
+public class ZOrderIntegers
+        implements ZOrder
+{
+    public static final int MAX_INPUT_DIMENSIONS = 10;
+
+    // Assume encodingBits is 1-based
+    private List<Integer> encodingBits;
+    private int totalBitLength;
+    private int maxBitLength;
+
+    public ZOrderIntegers(List<Integer> encodingBits)
+    {
+        this.encodingBits = encodingBits;
+
+        if (encodingBits.size() == 0) {
+            throw new IllegalArgumentException("Encoding bits list size should be greater than zero.");
+        }
+
+        totalBitLength = encodingBits.stream().mapToInt(Integer::intValue).sum();
+        maxBitLength = encodingBits.stream().mapToInt(Integer::intValue).max().getAsInt();
+    }
+
+    private void checkEncodeInputValidity(List<Integer> input)
+            throws IllegalArgumentException
+    {
+        if (input.size() == 0) {
+            throw new IllegalArgumentException("Input list size should be greater than zero.");
+        }
+
+        if (input.size() > MAX_INPUT_DIMENSIONS) {
+            throw new IllegalArgumentException(format("Our current Z-Ordering implementation does not support more than %d input numbers.", MAX_INPUT_DIMENSIONS));
+        }
+
+        if (input.size() > encodingBits.size()) {
+            throw new IllegalArgumentException(format("Input list size (%d) does not match encoding bits list size (%d).",
+                    input.size(), encodingBits.size()));
+        }
+
+        for (int index = 0; index < input.size(); index++) {
+            if (input.get(index) < 0) {
+                throw new IllegalArgumentException("Our current Z-Ordering implementation does not support negative input numbers.");
+            }
+        }
+    }
+
+    private void checkEncodeInputValidity(List<Integer> input, AddressType type)
+            throws IllegalArgumentException
+    {
+        checkEncodeInputValidity(input);
+
+        int maximumBits;
+        switch (type) {
+            case INT:
+                maximumBits = 32;
+                break;
+            case LONG:
+                maximumBits = 64;
+                break;
+            default:
+                throw new IllegalArgumentException("Our current Z-Ordering implementation only supports int and long address types.");
+        }
+
+        if (totalBitLength > maximumBits) {
+            throw new IllegalArgumentException(format("The z-address type specified is not large enough to hold %d bits.", totalBitLength));
+        }
+    }
+
+    private long convertByteAddressToLong(byte[] byteAddress)
+    {
+        long address = 0;
+        for (int byteIndex = 0; byteIndex < byteAddress.length; byteIndex++) {
+            address |= Byte.toUnsignedLong(byteAddress[byteIndex]) << (byteIndex << 3);
+        }
+        return address;
+    }
+
+    @Override
+    public byte[] encodeToByteArray(List input)
+            throws IllegalArgumentException
+    {
+        checkEncodeInputValidity(input);
```

Comment:
Do we still need this check here?

Reply:
Not sure if this has been resolved. For encodeToLong and encodeToInteger, checkEncodeInputValidity will be called twice. Let us keep this check here; meanwhile, (1) we should remove L70 (2) rename function name (L67) to "private static void checkEncodeOutputValidity(int totalEncodingBits, int expectedOutputBits)".